src/Tutorials/Custom-Grain-Storage.md

```diff
@@ -0,0 +1,232 @@
+            _grainFactory = grainFactory;
+            _typeResolver = typeResolver;
+        }
+
+        public Task ClearStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            throw new NotImplementedException();
+        }
+
+        public Task WriteStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+        {
+            throw new NotImplementedException();
+        }
+  
+        public void Participate(ISiloLifecycle lifecycle)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Participate(ISiloLifecycle lifecycle)
+        {
+            throw new NotImplementedException();
+        }
+    }
+}
+```
+
+We create a constructor containing two fields, `storageName` to specify which grains should write using this storage `[StorageProvider(ProviderName = "File")]` and `directory` which would be the directory where the grain states will be saved.
+
+`IGrainFactory`, `ITypeResolver` and `JsonSerializerSettings` will be used in the next section where we will initilize the storage.
+
+## Initializing the storage
+
+To initialize the storage, we register an `Init` function on the `ApplicationServices` lifecycle.
+
+```csharp
+public void Participate(ISiloLifecycle lifecycle)
+{
+    lifecycle.Subscribe(OptionFormattingUtilities.Name<FileGrainStorage>(_storageName), ServiceLifecycleStage.ApplicationServices, Init);
+}
+```
+
+The `Init` function is used to set the `_jsonSettings` which will be used to configure the `Json` serializer. At the same time we create the folder to store the grains states if it does not exist yet.
+
+```csharp
+private Task Init(CancellationToken ct)
+{
+    // Settings could be made configurable from Options.
+    _jsonSettings = OrleansJsonSerializer.UpdateSerializerSettings(OrleansJsonSerializer.GetDefaultSerializerSettings(_typeResolver, _grainFactory), false, false, null);
+
+    var directory = new System.IO.DirectoryInfo(_rootDirectory);
+    if (!directory.Exists)
+        directory.Create();
+
+    return Task.CompletedTask;
+}
+```
+
+## Reading State
+
+To read the state, to ensure uniqueness, we construct the filename using the grain key and the grain type.
+
+```csharp
+public async Task ReadStateAsync(string grainType, GrainReference grainReference, IGrainState grainState)
+{
+    var fName = grainReference.ToKeyString() + "." + grainType;
+    var path = Path.Combine(_rootDirectory, fName);
+
+    var fileInfo = new FileInfo(path);
+    if (!fileInfo.Exists)
+        return;
+
+    using (var stream = fileInfo.OpenText())
+    {
+        var storedData = await stream.ReadToEndAsync();
+        grainState.State = JsonConvert.DeserializeObject(storedData, _jsonSettings);
```

Comment:
If the implementation does not set or check the grainState.Etag, data loss is possible.  In case of Etag violation an InconsistentStateException (or some specialization thereof) needs be thrown.

Reply:
Thanks for the comment, that's interesting!

> If the implementation does not set or check the grainState.Etag, data loss is possible
Would you have an example to share where data loss would occur due to Etag not being checked?

> In case of Etag violation an InconsistentStateException (or some specialization thereof) needs be thrown.
I didn't see what the ETag was used for in the current AWS Dynamo and Azure Blob grain storages. Would you be able to guide me on where to look at?

Reply:
@Kimserey It varies a bit by implementation, I suppose. In ADO.NET it's checked like https://github.com/dotnet/orleans/blob/master/src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/AdoNetGrainStorage.cs#L193 . I believe all providers have not traditionally checked and thrown `InconsistentStateException` explicitly (I think ATS added check last winter).

Reply:
Note also https://github.com/dotnet/orleans/blob/master/src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/AdoNetGrainStorage.cs#L27 .

Reply:
>Would you have an example to share where data loss would occur due to Etag not being checked?

In the (rare) case of duplicate activations, one each grain would be able to load a modify and save the grain state and possibly overwrite the state changes of the other.

Assume A and B are the same grain, where B is a duplicate activation, caused by network partitioning, upgrade, or some other rare occurrence.

1) A loads state.
2) B loads state.
3) A modifies and saves state.
4) B modifies and saves state.

In the above, without etag validation the changes in 4) by B would overwrite the changes made by A in 3).  The expected behavior in the above would be for 4) to throw an InconsistentStateException, causing B to deactivate.

As you can probably see from looking through the grain storage implementations, they are inconsistent in many ways.  I won't go into the history of how we got to this point, but your document should help future implementations be more consistent in regards to the expected behaviors.

Reply:
Thanks @veikkoeeva and @jason-bragg, your comments helped be understand what was needed in term of ETag checks.

I added more information around ETag handling and also modified the code sections for `Read`, `Write` and `Clear`.

The behaviors are as followed:

- When grain reads state, ETag is set to last write date UTC
-  When grain writes state, current ETag is compared to last write date UTC of the file, if not equal, it means that a duplicate grain activation has written data in the state. So it throws exception making the activation kill itself
- When grain clears state, current ETag is compared to last write date UTC for the same reason as write

The previous behaviors also cater for non-existent files.

Thanks!

Reply:
@Kimserey Maybe add an explanation that in real operation the time resolution isn't sufficient and there's issue with locking the file, but this is just an example. Otherwise someone probably takes and uses this as-is. :)

Reply:
Thanks @veikkoeeva, I have added a stronger warning including the points you mentioned.